File: bcs-services/cluster-resources/pkg/action/util/resp/resp.go
/*
 * Tencent is pleased to support the open source community by making Blueking Container Service available.
 * Copyright (C) 2022 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * 	http://opensource.org/licenses/MIT
 *
 * Unless required by applicable law or agreed to in writing, software distributed under,
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */

package resp

import (
	"context"
	"strings"

	"google.golang.org/protobuf/types/known/structpb"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"

	"github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/common/errcode"
	res "github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/resource"
	cli "github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/resource/client"
	"github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/util/errorx"
	"github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/util/mapx"
	"github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/util/pbstruct"
)

// BuildListAPIResp ...
func BuildListAPIResp(
	ctx context.Context, clusterID, resKind, groupVersion, namespace, format string, opts metav1.ListOptions,
) (*structpb.Struct, error) {
	clusterConf := res.NewClusterConfig(clusterID)
	k8sRes, err := res.GetGroupVersionResource(ctx, clusterConf, resKind, groupVersion)
	if err != nil {
		return nil, err
	}

	var ret *unstructured.UnstructuredList
	ret, err = cli.NewResClient(clusterConf, k8sRes).List(ctx, namespace, opts)
	if err != nil {
		return nil, err
	}

	respDataBuilder, err := NewRespDataBuilder(ret.UnstructuredContent(), resKind, format)
	if err != nil {
		return nil, err
	}
	respData, err := respDataBuilder.BuildList()
	if err != nil {
		return nil, err
	}
	return pbstruct.Map2pbStruct(respData)
}

// BuildRetrieveAPIResp ...
func BuildRetrieveAPIResp(
	ctx context.Context, clusterID, resKind, groupVersion, namespace, name, format string, opts metav1.GetOptions,
) (*structpb.Struct, error) {
	clusterConf := res.NewClusterConfig(clusterID)
	k8sRes, err := res.GetGroupVersionResource(ctx, clusterConf, resKind, groupVersion)
	if err != nil {
		return nil, err
	}

	var ret *unstructured.Unstructured
	ret, err = cli.NewResClient(clusterConf, k8sRes).Get(ctx, namespace, name, opts)
	if err != nil {
		return nil, err
	}

	respDataBuilder, err := NewRespDataBuilder(ret.UnstructuredContent(), resKind, format)
	if err != nil {
		return nil, err
	}
	respData, err := respDataBuilder.Build()
	if err != nil {
		return nil, err
	}
	return pbstruct.Map2pbStruct(respData)
}

// BuildCreateAPIResp ...
func BuildCreateAPIResp(
	ctx context.Context, clusterID, resKind, groupVersion string, manifest map[string]interface{}, isNSScoped bool, opts metav1.CreateOptions,
) (*structpb.Struct, error) {
	clusterConf := res.NewClusterConfig(clusterID)
	k8sRes, err := res.GetGroupVersionResource(ctx, clusterConf, resKind, groupVersion)
	if err != nil {
		return nil, err
	}

	var ret *unstructured.Unstructured
	ret, err = cli.NewResClient(clusterConf, k8sRes).Create(ctx, manifest, isNSScoped, opts)
	if err != nil {
		return nil, err
	}
	return pbstruct.Unstructured2pbStruct(ret), nil
}

// BuildUpdateAPIResp ...
func BuildUpdateAPIResp(
	ctx context.Context, clusterID, resKind, groupVersion, namespace, name string, manifest map[string]interface{}, opts metav1.UpdateOptions,
) (*structpb.Struct, error) {
	clusterConf := res.NewClusterConfig(clusterID)
	k8sRes, err := res.GetGroupVersionResource(ctx, clusterConf, resKind, groupVersion)
	if err != nil {
		return nil, err
	}

	var ret *unstructured.Unstructured
	ret, err = cli.NewResClient(clusterConf, k8sRes).Update(ctx, namespace, name, manifest, opts)
	if err != nil {
		return nil, err
	}
	return pbstruct.Unstructured2pbStruct(ret), nil
}

// BuildDeleteAPIResp ...
func BuildDeleteAPIResp(
	ctx context.Context, clusterID, resKind, groupVersion, namespace, name string, opts metav1.DeleteOptions,
) error {
	clusterConf := res.NewClusterConfig(clusterID)
	k8sRes, err := res.GetGroupVersionResource(ctx, clusterConf, resKind, groupVersion)
	if err != nil {
		return err
	}
	return cli.NewResClient(clusterConf, k8sRes).Delete(ctx, namespace, name, opts)
}

// BuildListPodRelatedResResp ...
func BuildListPodRelatedResResp(
	ctx context.Context, clusterID, namespace, podName, format, resKind string,
) (*structpb.Struct, error) {
	relatedRes, err := cli.NewPodCliByClusterID(ctx, clusterID).ListPodRelatedRes(ctx, namespace, podName, resKind)
	if err != nil {
		return nil, err
	}
	respDataBuilder, err := NewRespDataBuilder(relatedRes, resKind, format)
	if err != nil {
		return nil, err
	}
	respData, err := respDataBuilder.BuildList()
	if err != nil {
		return nil, err
	}
	return pbstruct.Map2pbStruct(respData)
}

// BuildListContainerAPIResp ...
func BuildListContainerAPIResp(ctx context.Context, clusterID, namespace, podName string) (*structpb.ListValue, error) {
	podManifest, err := cli.NewPodCliByClusterID(ctx, clusterID).GetManifest(ctx, namespace, podName)
	if err != nil {
		return nil, err
	}

	containers := []map[string]interface{}{}
	containerStatuses, _ := mapx.GetItems(podManifest, "status.containerStatuses")
	for _, containerStatus := range containerStatuses.([]interface{}) {
		cs, _ := containerStatus.(map[string]interface{})
		status, reason, message := "", "", ""
		// state 有且只有一对键值：running / terminated / waiting
		// https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#containerstate-v1-core
		for k := range cs["state"].(map[string]interface{}) {
			status = k
			reason, _ = mapx.Get(cs, []string{"state", k, "reason"}, k).(string)
			message, _ = mapx.Get(cs, []string{"state", k, "message"}, k).(string)
		}
		containers = append(containers, map[string]interface{}{
			"containerID": extractContainerID(mapx.GetStr(cs, "containerID")),
			"image":       cs["image"].(string),
			"name":        cs["name"].(string),
			"status":      status,
			"reason":      reason,
			"message":     message,
		})
	}
	return pbstruct.MapSlice2ListValue(containers)
}

// BuildGetContainerAPIResp ...
func BuildGetContainerAPIResp(ctx context.Context, clusterID, namespace, podName, containerName string) (*structpb.Struct, error) {
	podManifest, err := cli.NewPodCliByClusterID(ctx, clusterID).GetManifest(ctx, namespace, podName)
	if err != nil {
		return nil, err
	}

	// 遍历查找指定容器的 Spec 及 Status，若其中某项不存在，则抛出错误
	var curContainerSpec, curContainerStatus map[string]interface{}
	containerSpec, _ := mapx.GetItems(podManifest, "spec.containers")
	for _, csp := range containerSpec.([]interface{}) {
		spec, _ := csp.(map[string]interface{})
		if containerName == spec["name"].(string) {
			curContainerSpec = spec
		}
	}
	containerStatuses, _ := mapx.GetItems(podManifest, "status.containerStatuses")
	for _, containerStatus := range containerStatuses.([]interface{}) {
		cs, _ := containerStatus.(map[string]interface{})
		if containerName == cs["name"].(string) {
			curContainerStatus = cs
		}
	}
	if len(curContainerSpec) == 0 || len(curContainerStatus) == 0 {
		return nil, errorx.New(errcode.General, "container %s spec or status not found", containerName)
	}

	// 各项容器数据组装
	containerInfo := map[string]interface{}{
		"hostName":      mapx.Get(podManifest, "spec.nodeName", "--"),
		"hostIP":        mapx.Get(podManifest, "status.hostIP", "--"),
		"containerIP":   mapx.Get(podManifest, "status.podIP", "--"),
		"containerID":   extractContainerID(mapx.GetStr(curContainerStatus, "containerID")),
		"containerName": containerName,
		"image":         mapx.Get(curContainerStatus, "image", "--"),
		"networkMode":   mapx.Get(podManifest, "spec.dnsPolicy", "--"),
		"ports":         mapx.Get(curContainerSpec, "ports", []interface{}{}),
		"volumes":       []map[string]interface{}{},
		"resources":     mapx.Get(curContainerSpec, "resources", map[string]interface{}{}),
		"command": map[string]interface{}{
			"command": mapx.Get(curContainerSpec, "command", []string{}),
			"args":    mapx.Get(curContainerSpec, "args", []string{}),
		},
	}
	mounts := mapx.Get(curContainerSpec, "volumeMounts", []map[string]interface{}{})
	for _, mount := range mounts.([]interface{}) {
		m, _ := mount.(map[string]interface{})
		containerInfo["volumes"] = append(containerInfo["volumes"].([]map[string]interface{}), map[string]interface{}{
			"name":      mapx.Get(m, "name", "--"),
			"mountPath": mapx.Get(m, "mountPath", "--"),
			"readonly":  mapx.Get(m, "readOnly", "--"),
		})
	}

	return pbstruct.Map2pbStruct(containerInfo)
}

// BuildUpdateCObjAPIResp 构建更新自定义资源请求响应结果
func BuildUpdateCObjAPIResp(
	ctx context.Context, clusterID, resKind, groupVersion, namespace, name string, manifest map[string]interface{}, opts metav1.UpdateOptions,
) (*structpb.Struct, error) {
	clusterConf := res.NewClusterConfig(clusterID)
	cobjRes, err := res.GetGroupVersionResource(ctx, clusterConf, resKind, groupVersion)
	if err != nil {
		return nil, err
	}

	// CustomObject 需要自行更新到最新的 ResourceVersion，否则会更新失败
	cobjManifest, err := cli.GetCObjManifest(ctx, clusterConf, cobjRes, namespace, name)
	if err != nil {
		return nil, err
	}
	latestRV, err := mapx.GetItems(cobjManifest, "metadata.resourceVersion")
	if err != nil {
		return nil, err
	}
	err = mapx.SetItems(manifest, "metadata.resourceVersion", latestRV)
	if err != nil {
		return nil, err
	}

	// 下发更新指令到集群
	var ret *unstructured.Unstructured
	ret, err = cli.NewResClient(clusterConf, cobjRes).Update(ctx, namespace, name, manifest, opts)
	if err != nil {
		return nil, err
	}
	return pbstruct.Unstructured2pbStruct(ret), nil
}

// 去除容器 ID 前缀，原格式：docker://[a-zA-Z0-9]{64}
func extractContainerID(rawContainerID string) string {
	return strings.Replace(rawContainerID, "docker://", "", 1)
}

File: bcs-ui/backend/dashboard/workloads/utils/resp.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
Edition) available.
Copyright (C) 2017-2021 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://opensource.org/licenses/MIT

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""
from typing import Dict, List

from attr import dataclass
from django.utils.translation import ugettext_lazy as _

from backend.dashboard.exceptions import ResourceNotExist
from backend.utils.basic import get_with_placeholder, getitems


@dataclass
class ContainerRespBuilder:
    """通过 Pod 配置信息获取容器信息"""

    pod: Dict
    container_name: str = None

    def build_list(self) -> List[Dict]:
        """构造列表展示的容器信息"""
        containers = []
        for cs in getitems(self.pod, 'status.containerStatuses', []):
            status = message = reason = None
            # state 有且只有一对键值：running / terminated / waiting
            # https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#containerstate-v1-core
            for k, v in cs['state'].items():
                status = k
                message = v.get('message') or k
                reason = v.get('reason') or k

            containers.append(
                {
                    # 原格式：docker://[a-zA-Z0-9]{64}，需要去除前缀，特殊状态下可能不存在
                    'container_id': self._extract_container_id(get_with_placeholder(cs, 'containerID')),
                    'image': cs['image'],
                    'name': cs['name'],
                    'status': status,
                    'message': message,
                    'reason': reason,
                }
            )
        return containers

    def build(self) -> Dict:
        """构造展示用的容器详情信息"""
        for cs in getitems(self.pod, 'status.containerStatuses', []):
            if self.container_name == cs['name']:
                container_status = cs
                break
        else:
            raise ResourceNotExist(_('容器 {} 状态信息不存在').format(self.container_name))

        labels = getitems(self.pod, 'metadata.labels', {})
        spec, status = self.pod['spec'], self.pod['status']
        for csp in spec.get('containers', []):
            if self.container_name == csp['name']:
                container_spec = csp
                break
        else:
            raise ResourceNotExist(_('容器 {} 模板（Spec）信息不存在').format(self.container_name))

        return {
            'host_name': get_with_placeholder(spec, 'nodeName'),
            'host_ip': get_with_placeholder(status, 'hostIP'),
            'container_ip': get_with_placeholder(status, 'podIP'),
            'container_id': self._extract_container_id(get_with_placeholder(container_status, 'containerID')),
            'container_name': self.container_name,
            'image': get_with_placeholder(container_status, 'image'),
            'network_mode': get_with_placeholder(spec, 'dnsPolicy'),
            # 端口映射
            'ports': container_spec.get('ports', []),
            # 命令
            'command': {
                'command': get_with_placeholder(container_spec, 'command', ''),
                'args': ' '.join(container_spec.get('args', [])),
            },
            # 挂载卷
            'volumes': [
                {
                    'host_path': get_with_placeholder(mount, 'name'),
                    'mount_path': get_with_placeholder(mount, 'mountPath'),
                    'readonly': get_with_placeholder(mount, 'readOnly'),
                }
                for mount in container_spec.get('volumeMounts', [])
            ],
            # 标签
            'labels': [{'key': key, 'val': val} for key, val in labels.items()],
            # 资源限制
            'resources': container_spec.get('resources', {}),
        }

    def _extract_container_id(self, container_id: str) -> str:
        """
        去除 容器 ID 前缀，原格式：docker://[a-zA-Z0-9]{64}

        :param container_id: 原始的容器 ID
        :return: 去除前缀后的容器 ID
        """
        return container_id.replace('docker://', '')
